### ISSUE TITLE
SyncCommandBufferBuilderError when reading from an AttachmentImage twice in one render pass

### ISSUE THREAD

**Issue by IcyDefiance** [vulkano-rs/vulkano#1234]
*   Version of vulkano: master branch
*   OS: Windows 10
*   GPU (the selected PhysicalDevice): GTX 1080 Ti

# What I'm doing

1. Create an AttachmentImage
2. Initialize it with clear_color_image, then set a fence and wait on it
3. Use it as a texture in two different descriptor sets in the same renderpass

# What vulkano is doing

1. When binding the first descriptor set, it hits [base.rs:655](https://github.com/vulkano-rs/vulkano/blob/814dacdab923771f8a818f7925785d347ddd4866/vulkano/src/command_buffer/synced/base.rs#L655).
    * initial_layout_requirement == ColorAttachmentOptimal
    * start_layout == ShaderReadOnlyOptimal
2. When binding the second descriptor set, it hits [base.rs:533](https://github.com/vulkano-rs/vulkano/blob/814dacdab923771f8a818f7925785d347ddd4866/vulkano/src/command_buffer/synced/base.rs#L533).
    * entry.exclusive == true

---

I deleted [base.rs:655](https://github.com/vulkano-rs/vulkano/blob/814dacdab923771f8a818f7925785d347ddd4866/vulkano/src/command_buffer/synced/base.rs#L655) on my fork and it solved the problem for me, but I'm guessing that also disables some important validation.

**Comment by AustinJ235** [comment]
I believe this is intentional. You shouldn't be able to write and read within the same renderpass. Although this does work on some gpus this is against spec.

**Comment by IcyDefiance** [comment]
Sorry for the delayed response. I needed to look into this again to be sure about what was happening.

The problem is that I'm not writing to it at all during the renderpass. I'm only writing to it with clear_color_image, before the renderpass starts. (In fact, clear_color_image can't be called at all inside a renderpass.)

Vulkano then lets me read from the image once during the renderpass, but if I try to read from it a second time, it returns the SyncCommandBufferBuilderError.

Looking at it now, with just a little more knowledge than I had 2 months ago, I really don't think the `actually_exclusive = true;` line should be there. It doesn't make sense that adding a memory barrier would prevent you from accessing the image again until the renderpass ends. I'm just not certain enough to make a PR until someone who knows more about Vulkan figures out what is going on.

**Comment by AustinJ235** [comment]
if you write at all during the renderpass excluding the begin renderpass, you are unable to read. You'll have to use another pass.

**Comment by IcyDefiance** [comment]
Does the memory barrier itself count as writing, then? If so, then why can I still read from the image once?

**Comment by IcyDefiance** [comment]
Here's a minimal demonstration: https://github.com/IcyDefiance/target-texture-example

I just copied the image example from this repo and modified it to use an AttachmentImage that is initialized with `clear_color_image()`. Here's the diff: https://www.diffchecker.com/rnYluVi2

As it is, it will panic on the first frame, but if you comment out the second draw call (line 204) then it will work perfectly.

I also noticed that the error only happens when using two different descriptor sets. If you make the second draw call use `set` instead of `set2`, it won't panic. I'm not sure yet why this is.
